### PR TITLE
Add `cSpell` to NUKE build pipeline

### PR DESCRIPTION
**Pull request by IT-VBFK**
This ref. and finishes #2017 

## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [ ] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [ ] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [ ] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [ ] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
  * [x] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

File: .github/workflows/spellcheck.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    paths:
+      - "docs/**/*"
+jobs:
+  spellcheck:
+    name: "Docs: Spellcheck"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        name: Check out the code
+      - uses: actions/setup-node@v1
```
Comment (jnyrup) [inline .github/workflows/spellcheck.yml:22]:
```suggestion
      - uses: actions/setup-node@v3
```

File: .github/workflows/spellcheck.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    branches:
+      - develop
+    paths:
+      - "docs/**/*"
+jobs:
+  spellcheck:
+    name: "Docs: Spellcheck"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
[the remark this answers is not in the document]
Comment (IT-VBFK) [reply .github/workflows/spellcheck.yml]: Hmm I've seen this before. Back then setting the fetch depth to 0 (meaning disable a shallow clone) fixes this issue

**Comment by IT-VBFK** [comment]
So.. now I cannot imagine how to move more steps to .NET 😆

File: Build/Build.cs [hunk 3]
```diff
@@ -252,27 +252,20 @@ class Build : NukeBuild
         });
 
     Target Spellcheck => _ => _
+        .OnlyWhenDynamic(() => !IsLocalBuild)
```
[the remark this answers is not in the document]
Comment (IT-VBFK) [reply Build/Build.cs]: Solved by adding Node.js.redist package to the build project

File: .github/workflows/spellcheck.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+name: Documentation Checks
```
Comment (dennisdoomen) [inline .github/workflows/spellcheck.yml:1]: 🤔 Why do we need a separate build for this? Why not make the `SpellCheck` target part of the normal build?

File: Build/Build.cs [hunk 5]
```diff
@@ -246,5 +258,22 @@ class Build : NukeBuild
...
         });
 
+    Target Spellcheck => _ => _
+        .Executes(() =>
+        {
+            InstallCSpell();
+            RunCSpell();
+        });
+
+    private void InstallCSpell()
```
Comment (dennisdoomen) [inline Build/Build.cs:268]: ⛏️ I prefer to not have small methods which name literally repeats what the single line is doing.

File: Build/Build.cs [hunk 6]
```diff
@@ -246,5 +258,22 @@ class Build : NukeBuild
...
+    {
+        Node($"{YarnCli} install --silent", workingDirectory: RootDirectory);
+    }
+
+    private void RunCSpell()
+    {
+        Node($"{YarnCli} run cspell");
```
Comment (dennisdoomen) [inline Build/Build.cs:275]: 🔧 Node always writes its output to the wrong channel, resulting in "errors" like below. That's why this `LogErrorAsWarning` custom logger was needed in the example. 

image

Comment (jnyrup) [inline Build/Build.cs:275]: image

https://github.com/fluentassertions/fluentassertions/actions/runs/3845958130/jobs/6550706354

Is there some exit code confusion between cSpell and NUKE 🤔

**Review by dennisdoomen — CHANGES_REQUESTED** [review]

File: Build/Build.cs (hunk 6 above)
Comment (IT-VBFK) [reply Build/Build.cs:275]: Yes.. the reason for this is: cspell sends this to stderr. And NUKE takes this and prints an error.

I had this sort of "error" several times in my day job. (Was building docker images with NUKE).

I haven't found a solution yet

File: .github/workflows/spellcheck.yml (hunk 4 above)
Comment (IT-VBFK) [reply .github/workflows/spellcheck.yml:1]: Now, the build.yml ignores changes inside the docs directory. If you are good to run a whole build process when only updating the docs, I will merge this to files together.

File: Build/Build.cs (hunk 5 above)
Comment (IT-VBFK) [reply Build/Build.cs:268]: Ok

File: Build/Build.cs (hunk 6 above)
Comment (IT-VBFK) [reply Build/Build.cs:275]: Ahh.. now I get it. Will change that in a second

Comment (IT-VBFK) [reply Build/Build.cs:275]: This idea of writing stderr only as warning is quite cool.. learned something for future 👍🏻 Thanks

Comment (IT-VBFK) [reply Build/Build.cs:275]: Fixed :)

File: cSpell.json [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        "formattable",
+        "Guids",
+        "LINQ",
+        "parameterless",
+        "refactorings",
+        "struct"
+    ],
+    "ignoreWords": [
+        "Gallio"
+    ],
```
Comment (jnyrup) [inline cSpell.json:16]: is this still needed when `docs/_pages/releases.md` is ignored?

File: .github/workflows/spellcheck.yml (hunk 4 above)
Comment (jnyrup) [reply .github/workflows/spellcheck.yml:1]: I think I like having it as a separate build (but I'm no expert in all the pros and cons here)

Benefits of having it a separate build
* no need to run spellcheck for all three operating systems
* no need to run spell check when not modifying docs
* no need to run regular unit-tests when only modifying docs

File: cSpell.json (hunk 7 above)
Comment (IT-VBFK) [reply cSpell.json]: True

File: .github/workflows/spellcheck.yml (hunk 4 above)
Comment (dennisdoomen) [reply .github/workflows/spellcheck.yml:1]: The biggest con is that you will only get the spell checking issues after your PR is build. On the other hand, I guess most people won't run the local build script anyway. So maybe keeping it separate is fine after all. 

And maybe we can invoke it as part of the `run.bat` in the `docs` folder?

Comment (IT-VBFK) [reply .github/workflows/spellcheck.yml:1]: Sounds good. 

Before or after running jekyll? Or does this matter in any way?

Comment (dennisdoomen) [reply .github/workflows/spellcheck.yml:1]: > What if we add a note in the contributing guidelines to run `build.(sh|ps1) --target spellcheck` before pushing (if the docs are effected)

We should most definitely do that. 

> Before or after running jekyll? Or does this matter in any way?

I guess doing this before building the site makes more sense. And `run.bat` is blocking.

Comment (IT-VBFK) [reply .github/workflows/spellcheck.yml:1]: Ok.. added both :)

**Review by dennisdoomen — APPROVED** [review]
Nice!

File: Build/Build.cs [hunk 8]
```diff
@@ -246,5 +259,25 @@ class Build : NukeBuild
...
+            static void LogErrorAsWarning(OutputType outputType, string message)
+            {
+                if (outputType == OutputType.Err)
+                {
+                    Warning(message);
+                }
+                else
+                {
+                    Information(message);
+                }
```
Comment (IT-VBFK) [inline Build/Build.cs:274]: I have noticed that this construct prints warnings as information.. Should we change that?

Comment (dennisdoomen) [reply Build/Build.cs:274]: Hmm, good point. I haven't even noticed the real warnings among all the others ;-)

Comment (IT-VBFK) [reply Build/Build.cs:274]: Ahh.. Nevermind.. `OutputType.Err` means output from stderr.

**Comment by jnyrup** [comment]
I had a play with this, and here's my best attempt

```cs
Node($"{YarnCli} install --silent", workingDirectory: RootDirectory);
Node($"{YarnCli} --silent run cspell --no-summary", workingDirectory: RootDirectory,
    customLogger: (_, msg) => Error(msg));
```

Without spelling errors
image

With spelling errors
image

**Comment by IT-VBFK** [comment]
That looks great.. Just changed it :)

**Comment by dennisdoomen** [comment]
> I had a play with this, and here's my best attempt

What was the important trick? The `--silent --no-summary`?

**Comment by jnyrup** [comment]
> > I had a play with this, and here's my best attempt
> 
> What was the important trick? The `--silent --no-summary`?

At a second look, only the `--no-summary` is necessary as it prevents `cspell` from _always_ printing a summary to stderr.
(This is what happens when messing with technologies without knowing them 🤷 )

So this seems sufficient:
```cs
Node($"{YarnCli} install --silent", workingDirectory: RootDirectory);
Node($"{YarnCli} run cspell --no-summary", workingDirectory: RootDirectory);
```

image


image

File: Build/Build.cs [hunk 9]
```diff
@@ -246,5 +259,14 @@ class Build : NukeBuild
...
 
+    Target SpellCheck => _ => _
+        .Executes(() =>
+        {
+            Node($"{YarnCli} install --silent", workingDirectory: RootDirectory);
+            Node($"{YarnCli} --silent run cspell --no-summary", workingDirectory: RootDirectory, 
+                customLogger: (_, msg) => Error(msg));
+        });
+
+
```
Comment (jnyrup) [inline Build/Build.cs:270]:
```suggestion

```

File: Build/Build.cs [hunk 10]
```diff
@@ -47,11 +48,23 @@ class Build : NukeBuild
...
+#if OS_WINDOWS
+    [PackageExecutable("Node.js.redist", "node.exe", Version = "16.17.1", Framework = "win-x64")]
+    Tool Node;
+#elif OS_MAC
+    [PackageExecutable("Node.js.redist", "node", Version = "16.17.1", Framework = "osx-x64")]
+    Tool Node;
+#else
+    [PackageExecutable("Node.js.redist", "node", Version = "16.17.1", Framework = "linux-x64")]
+    Tool Node;
+#endif
```
Comment (jnyrup) [inline Build/Build.cs:60]:
```suggestion
#if OS_WINDOWS
    const string NodeFramework = "win-x64";
#elif OS_MAC
    const string NodeFramework = "osx-x64";
#else
    const string NodeFramework = "linux-x64";
#endif

    [PackageExecutable("Node.js.redist", "node", Version = "16.17.1", Framework = NodeFramework)]
    Tool Node;
```

**Comment by IT-VBFK** [comment]
But now you cannot see the misspelled words in the summary

**Comment by IT-VBFK** [comment]
Sorry for the bad commit messages I am on my mobile device right now 🫣

**Comment by jnyrup** [comment]
Third attempt...
Here are screenshots of what happens with each extra argument.
I guess(?) the third snippet here produces the most desired output.

```cs
Node($"{YarnCli} run cspell --no-summary", workingDirectory: RootDirectory);
```
image

```cs
Node($"{YarnCli} run cspell --no-summary", workingDirectory: RootDirectory,
    customLogger: (_, msg) => Error(msg));
```
image

```cs
Node($"{YarnCli} --silent run cspell --no-summary", workingDirectory: RootDirectory,
    customLogger: (_, msg) => Error(msg));
```
image

File: Build/Build.cs [hunk 11]
```diff
@@ -47,11 +48,23 @@ class Build : NukeBuild
...
+#if OS_WINDOWS
+    const string NodeFramework = "win-x64";
+#elif OS_MAC
+    const string NodeFramework = "osx-x64";
+#else
+    const string NodeFramework = "linux-x64";
+#endif
+
+    [PackageExecutable("Node.js.redist", "node", Version = "16.17.1", Framework = NodeFramework)]
+    Tool Node;
```
Comment (jnyrup) [inline Build/Build.cs:60]: I failed to see that windows requires `node.exe`

Now I beginning to doubt a bit whether this becomes less readable than having the constants inlined
@dennisdoomen opinions?

```suggestion
#if OS_WINDOWS
    const string NodeFramework = "win-x64";
    const string NodeExecutable = "node.exe";
#elif OS_MAC
    const string NodeFramework = "osx-x64";
    const string NodeExecutable = "node";
#else
    const string NodeFramework = "linux-x64";
    const string NodeExecutable = "node";
#endif

    [PackageExecutable("Node.js.redist", NodeExecutable, Version = "16.17.1", Framework = NodeFramework)]
    Tool Node;
```

**Comment by IT-VBFK** [comment]
I have meant this:

image

I think this is the simplest solution (#3)

**Comment by IT-VBFK** [comment]
I think wrapping only the attribute into preprocessor directives reads less noisy.. 

What do you think @jnyrup @dennisdoomen ?

File: Build/Build.cs (hunk 11 above)
Comment (dennisdoomen) [reply Build/Build.cs]: Defining them separately actually makes it _less_ readable. Just in-line them please...

Comment (IT-VBFK) [reply Build/Build.cs]: Like my latest commit?

**Review by jnyrup — APPROVED** [review]

**Comment by IT-VBFK** [comment]
@jnyrup or @dennisdoomen Do you mind re-running the pipeline on develop?

Thanks

**Comment by jnyrup** [comment]
I'm re-running now

**Comment by IT-VBFK** [comment]
Thanks.. this little red X driving me nuts :D

**Comment by SeanKilleen** [comment]
Hey, I was going through old PRs etc. and I just wanted to say, I love that you followed through on this, @IT-VBFK! I got to read it and learn more about Nuke and how you accomplished it. Thanks!

**Pull request merged**